packages/backend/src/api/controllers/FrontendController.ts

```diff
@@ -355,4 +363,99 @@ export class FrontendController {
       status: 200,
     }
   }
+
+  async getSearchRedirect(query: string): Promise<ControllerResult> {
+    const queryType = categorizeSearchQuery(query)
+
+    switch (queryType) {
+      case 'state_root_hash_or_stark_key': {
+        return this.searchForRootHashOrStarkKey(query)
+      }
+
+      case 'ethereum_address':
+        return this.searchForEthereumAddress(query)
+
+      default:
+        return {
+          status: 404,
+          html: "Search query param couldn't be found",
+        }
+    }
+  }
+
+  private async searchForEthereumAddress(
+    ethereumAddr: string
+  ): Promise<ControllerResult> {
+    const userRegistrationEvent =
+      await this.userRegistrationEventRepository.findByEthereumAddr(
+        ethereumAddr
+      )
+    if (userRegistrationEvent === undefined) {
+      return {
+        status: 404,
+        html: 'Ethereum address not found!',
+      }
+    }
+
+    const positionId =
+      await this.stateUpdateRepository.getPositionIdByPublicKey(
+        userRegistrationEvent.starkKey
+      )
+
+    return {
+      status: 302,
+      url: `/positions/${positionId}`,
+    }
+  }
+
+  private async searchForRootHashOrStarkKey(
+    rootHashOrStarkKey: string
+  ): Promise<ControllerResult> {
+    const positionId = await this.stateUpdateRepository.getPositionIdByRootHash(
+      rootHashOrStarkKey.slice(2)
+    )
+    if (positionId) {
+      return {
+        status: 302,
+        url: `/state-updates/${positionId}`,
+      }
+    }
+
+    return await this.searchForStarkKey(rootHashOrStarkKey)
+  }
+
+  private async searchForStarkKey(starkKey: string): Promise<ControllerResult> {
+    const userRegistrationEvent =
+      await this.userRegistrationEventRepository.findByStarkKey(starkKey)
+
+    if (userRegistrationEvent === undefined) {
+      return {
+        status: 404,
+        html: 'State root hash not found!',
+      }
+    }
+
+    const positionId =
+      await this.stateUpdateRepository.getPositionIdByPublicKey(
+        userRegistrationEvent.starkKey
+      )
+
+    return {
+      status: 302,
+      url: `/positions/${positionId}`,
+    }
+  }
+}
+
+type Type = 'state_root_hash_or_stark_key' | 'ethereum_address' | 'unknown'
+
+function categorizeSearchQuery(query: string): Type {
```

Comment:
It could be easier if this method was names: `parseSearchQuery` and it returned a struct:
```ts
interface ParsedQuery {
  ethereumAddress: EthereumAddress | undefined,
  pedersenHash: PedersenHash | undefined,
  hash256: Hash256 | undefined,
  starkKey: string | undefined // TODO: type for StarkKey
  id: number | undefined
}
```
This way you preserve the static typing and enforce all potential constraints like ethereum address casing.
Then the code could just check for each option if it is undefined or not and try to return values accordingly.